### PR TITLE
feat(customer_v2): Add customer V2 delete api

### PR DESCRIPTION
**Pull request by sahkal**
## Type of Change

- [x] New feature

## Description

Add customer V2 delete api

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Create customer and delete a customer for v1 sanity

```
curl --location --request DELETE 'http://localhost:8080/customers/cus_7cGtbblt3oTLHQMqLget' \
--header 'Accept: application/json' \
--header 'api-key: dev_H6FAgVKbzAyNYgqjO6ZkISjNvR99LWKsmAP4TE3arzqebApYliPP5VaRhyusPxch'
```
Screenshot 2024-08-30 at 4 32 57 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code

### CODE REVIEW

File: crates/router/src/core/customers.rs [hunk 1]
```diff
@@ -462,174 +466,366 @@ pub async fn list_customers(
...
+                &key_store,
+                merchant_account.storage_scheme,
+            )
+            .await
+            .switch()?;
+
+        let merchant_reference_id = customer_orig.merchant_reference_id.clone();
+
+        let customer_mandates = db
+            .find_mandate_by_merchant_id_global_id(merchant_account.get_id(), &self.id)
```
Comment (jarnura) [inline crates/router/src/core/customers.rs:514]: we don't need merchant id in these types of queries.

File: crates/router/src/core/customers.rs [hunk 2]
```diff
@@ -462,174 +466,366 @@ pub async fn list_customers(
...
+            .switch()?;
+
+        for mandate in customer_mandates.into_iter() {
+            if mandate.mandate_status == enums::MandateStatus::Active {
+                Err(errors::CustomersErrorResponse::MandateActive)?
+            }
+        }
+
+        match db
+            .find_payment_method_by_global_id_merchant_id_list(
```
Comment (jarnura) [inline crates/router/src/core/customers.rs:525]: we don't need merchant id in these types of queries.

File: crates/router/src/core/customers.rs [hunk 3]
```diff
@@ -462,174 +466,366 @@ pub async fn list_customers(
...
-    req: customers::CustomerId,
+    req: customers::GlobalId,
     key_store: domain::MerchantKeyStore,
 ) -> errors::CustomerResponse<customers::CustomerDeleteResponse> {
-    let db = &state.store;
+    let db = &*state.store;
     let key_manager_state = &(&state).into();
-    let customer_orig = db
-        .find_customer_by_customer_id_merchant_id(
+    req.generate_delete_customer_response_fetch_domain_model_and_update(
```
Comment (jarnura) [inline crates/router/src/core/customers.rs:479]:
```suggestion
    req.fetch_domain_model_and_update_and_generate_delete_customer_response(
```

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payment_methods/cards.rs [hunk 4]
```diff
@@ -1978,6 +1978,31 @@ pub async fn delete_card_from_locker(
     .await
 }
 
+#[cfg(all(feature = "v2", feature = "customer_v2"))]
+pub async fn delete_card_from_locker_by_global_id(
```
Comment (jarnura) [inline crates/router/src/core/payment_methods/cards.rs:1982]: this should be by locker_id and not global_id

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payment_methods/cards.rs (hunk 4 above)
Comment (sahkal) [reply crates/router/src/core/payment_methods/cards.rs:1982]:
```
pub async fn delete_card_from_locker(
    state: &routes::SessionState,
    customer_id: &id_type::CustomerId,
    merchant_id: &id_type::MerchantId,
    card_reference: &str,
) -> errors::RouterResult<payment_methods::DeleteCardResp> {
    metrics::DELETE_FROM_LOCKER.add(&metrics::CONTEXT, 1, &[]);

    common_utils::metrics::utils::record_operation_time(
        async move {
            delete_card_from_hs_locker(state, customer_id, merchant_id, card_reference)
                .await
                .inspect_err(|error| {
                    metrics::CARD_LOCKER_FAILURES.add(&metrics::CONTEXT, 1, &[]);
                    logger::error!(?error, "Failed to delete card from locker");
                })
        },
        &metrics::CARD_DELETE_TIME,
        &metrics::CONTEXT,
        &[],
    )
    .await
}
```

but the v1 function of it doesnot delete it by locker_id! basically i think it's an api call to locker where we get locker info from the state

Comment (jarnura) [reply crates/router/src/core/payment_methods/cards.rs:1982]: This function itself not relevant to v2 just make it todo or deprecate it.

Comment (sahkal) [reply crates/router/src/core/payment_methods/cards.rs:1982]: okay!

**Review by jarnura — APPROVED** [review]

**Review by prajjwalkumar17 — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Review by prajjwalkumar17 — APPROVED** [review]

**Review by tsdk02 — APPROVED** [review]

**Pull request merged**